### PR TITLE
Add an RFC for OT-based synchronization.

### PR DESCRIPTION
**Pull request by djmitche**
This has been discussed in #2507 and several other venues.  I'd like to focus the proposal in one place.  I'm not sure if a PR containing raw HTML is a good idea, but it seems to be the only option here.

How could I best get a bunch of interested eyes on this proposal?

### CODE REVIEW

**Comment by tbabej** [comment]
> This has been discussed in #2507 and several other venues.  I'd like to focus the proposal in one place.  I'm not sure if a PR containing raw HTML is a good idea, but it seems to be the only option here.

This is long overdue, but in a couple of weeks the rewrite of this website to Markdown-based static generator should be finished. Assuming we're going to keep this PR open for a while to have a central place for the discussion, we could just use Markdown version of this RFC for now here and rebase the PR once the repository layout switches.

I was also thinking of moving parts of the docs to the taskwarrior repo itself. But that is a separate discussion.

File: html/docs/design/sync-ot.html [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+                change. Once finalized, the feature will be scheduled for an
+                upcoming release.
+              </p>
+            </div>
+
+            <a name="sync"></a>
+            <h3>Task Synchronization</h3>
+            <p>
+              This RFC covers <em>synchronization</em> of <em>replicas</em>
+              containing the same set of tasks.  A replica is can perform all
```
Comment (tbabej) [inline html/docs/design/sync-ot.html:69]:
```suggestion
              containing the same set of tasks.  A replica can perform all
```

File: html/docs/design/sync-ot.html [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+            <p>
+            All modifications to that state occur via <em>operations</em>.
+            We can draw a network, or graph, with the nodes representing states and the edges representing operations.
+            For example:
+            </p>
+<pre>  o -- State: {"abc-123": {"description": "get groceries", "priority": "L"}}
+  |
+  | -- Operation: set abc-123 "priority" to "H"
+  |
+  o -- State: {"abc-123": {"description": "get groceries", "priority: "H"}
```
Comment (tbabej) [inline html/docs/design/sync-ot.html:129]:
```suggestion
  o -- State: {"abc-123": {"description": "get groceries", "priority: "H"}}
```

File: html/docs/design/sync-ot.html [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+            The synchronization process uses operational transformation to "linearize" those operations.
+            This process is analogous (vaguely) to rebasing a sequence of Git commits.
+            </p>
+
+            <h4 id="versions">Versions</h4>
+
+            <p>
+            Occasionally, database states are given a name (that takes the form of a UUID).
+            The system as a whole (all replicas) constructs a branch-free sequence of versions and the operations that separate each version from the next.
+            The version with the nil UUID is implicitly the empty database.
```
Comment (tbabej) [inline html/docs/design/sync-ot.html:157]: Perhaps we should use a different term here, to avoid overloading of UUIDs? :thinking:

File: html/docs/design/sync-ot.html [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+              <li>
+                <p>
+                As designed, the sequence of versions grows without bound.  This consumes growing storage space on the server, and requires any new replicas to apply a lengthy sequence of operations contained in the versions since the first synchronization.
+                A "snapshot" approach may be useful here, allowing the server to discard versions which have been incorporated into snapshots, and allowing new replicas to download the latest snapshot and apply only the versions after that snapshot.
+                The heuristics of this approach need careful consideration.
+                </p>
+              </li>
+              <li>
+                <p>
+                Tasks must be deleted to avoid unbounded growth of the set of tasks.
```
Comment (tbabej) [inline html/docs/design/sync-ot.html:303]: Perhaps this could be made clearer that this is a goal that subset of users require, not something that this RFC imposes on all users?

Speaking from the personal perspective, I don't have interest in trimming my modification history. As outlined elsewhere, my task DB is not reaching sizes that would meaningfully slow the program down. The `undo.data`, which contains all the modifications (in a very wasteful format) is 3.5MiB of data. I suspect changing the data format here to just containing the operations themselves would lead to ~3-5x decrease in storage requirements.

TL;DR: I take certain nostalgic value in being able to see all the tasks with all the modifications ever performed on them :sweat_smile:

File: html/docs/design/sync-ot.html [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+                <p>
+                Synchronization becomes a required kind of periodic maintenance, even if against a local implementation.
+                This may be performed automatically, or existing users may be required to run <code>task sync</code> periodically.
+                </p>
+              </li>
+
+              <li>
+                <p>
+                The current on-disk storage for a replica is text-based and invites users to edit it.
+                This will interact particularly badly with operation-based synchronization, so it may be necessary to choose a more robust format.
```
Comment (tbabej) [inline html/docs/design/sync-ot.html:369]: I'd claim this is orthogonal, in fact, the current data storage format has exactly the same amount of danger. The data about the tasks is stored in `pending.data`, the task modifications (akin to the operations proposed to in this document are stored in `backlog.data`). Modifying `pending.data` manually and expecting sync to work would lead to the same data corruption as described here even with the current implementation.

I went over (triaged / reproduced where applicable) several hundred issues in the past year. For what it's worth, so far the number of bugs that I've encountered whose root problem was hand-editing the `.data` files was 0. While I support the exploration of alternative data stores as outlined in the discussion [about atomic data file writes](https://github.com/GothenburgBitFactory/taskwarrior/issues/152) I would say the evidence that this is something that users do is not there.

File: html/docs/design/sync-ot.html (hunk 3 above)
Comment (tbabej) [reply html/docs/design/sync-ot.html:157]: I see this is referred to as _version_ in latter parts of the text.

File: html/docs/design/sync-ot.html [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+                Without an understanding of tags, transposing these two operations requires one to "win", effectively undoing the other.
+                If, instead, each tag is stored as a distinct task key, then the two operations to add tags are <code>Update($uuid, "tag_bing", "")</code> and <code>Update($uuid, "tag_next", "")</code>.
+                These operations do not conflict with each other and can be transposed easily.
+                </p>
+              </li>
+
+              <li>
+                <p>
+                Synchronization becomes a required kind of periodic maintenance, even if against a local implementation.
+                This may be performed automatically, or existing users may be required to run <code>task sync</code> periodically.
```
Comment (tbabej) [inline html/docs/design/sync-ot.html:362]: > Synchronization becomes a required kind of periodic maintenance, even if against a local implementation. 
                 
It's a bit unclear what this paragraph is trying to say. Would users that don't use `sync` be required to run some dummy version of `sync`? I don't see evidence for that in the rest of the document, but I'm not sure how to interpret this.

**Comment by tbabej** [comment]
> How could I best get a bunch of interested eyes on this proposal?

There are couple of options:
* we could promote this RFC on the homepage
* ask for feedback in other TW communities (reddit, twitter, mailing list ..)

I can help with some of that, if you'd like. But first we should decide what is the process here. Are we going to postpone merging, and use this as the place to discuss (via review process) and only merge once finalized, or are we going to merge after first round of comments to be able to promote the website deployed version of this document?

File: html/docs/design/sync-ot.html (hunk 3 above)
Comment (djmitche) [reply html/docs/design/sync-ot.html:157]: The term is "version", yes -- the uuid is the format (although it doesn't have to be)

File: html/docs/design/sync-ot.html (hunk 4 above)
Comment (djmitche) [reply html/docs/design/sync-ot.html:303]: Good point.  This can certainly be configurable.

File: html/docs/design/sync-ot.html (hunk 5 above)
Comment (djmitche) [reply html/docs/design/sync-ot.html:369]: Ah, that's good data.  I'll remove this section.

File: html/docs/design/sync-ot.html (hunk 6 above)
Comment (djmitche) [reply html/docs/design/sync-ot.html:362]: That is an option, but we might also be able to do so automatically for cases where synchronization is not configured.  I'm trying to leave that open-ended here.

**Comment by djmitche** [comment]
That's a good question.  How has this been done in the past?  What level of interest do you expect?  PR comments are nice to a point, and beyond that point they are overwhelming.  There's also the question of format -- I actually translated markdown to html to create this document!

Maybe let's do two phases:
 1. You and others actively working on the project comment on this PR, mostly from a technical perspective (as you've done already).
 2. We merge that, and put it on tw.org, and then promote on reddit, twitter, etc. with a link to a discussion forum for the topic, and with a timeframe (2 weeks?) on those discussions after which I'll revise and -- assuming the revisions are minor -- check with the folks who brought up the points requiring revision.  If the revisions are major, we can repeat the process.

File: html/docs/design/sync-ot.html (hunk 2 above)
Comment (tbabej) [reply html/docs/design/sync-ot.html:129]: @djmitche This one is still outstanding in the current revision.

Comment (djmitche) [reply html/docs/design/sync-ot.html:129]: Oh, the "Commit Suggestion" button didn't work, oops.

**Comment by djmitche** [comment]
> This is long overdue, but in a couple of weeks the rewrite of this website to Markdown-based static generator should be finished. 

Apologies for my poor GitHub searching skills.  Can you point me to this effort?

**Comment by tbabej** [comment]
> Apologies for my poor GitHub searching skills. Can you point me to this effort?

No apologies needed, I haven't pushed it up until this very moment (have a look at the [`md-rewrite`](https://github.com/GothenburgBitFactory/tw.org/tree/md-rewrite) branch). The MD rewrite (very much WIP) might not even build for you, as there is a lot of uncomitted stuff that I still have locally (around theme and other parts).

That said, I pushed one of the converted old RFCs (the one for recurrence), so as long as you put in the YAML frontmatter (only the title is required) and put the file in the `content/docs/design/` folder, it will be compatible with the rest of the work.

**Comment by djmitche** [comment]
Ah. very cool.  I will wait for that to land before rewriting -- I don't think there's any immediate rush.

**Comment by tbabej** [comment]
As discussed with Dustin elsewhere, we could use an approach along the following lines to utilize a append-only operations log that would replace both `undo.data` and `backlog.data`:

Operations could have, alongside the regular content (modification, uuid and such) two additional attributes:

* the "action uuid", where operations generated by the same user action (i.e. bulk modification) would share an identifier
* the "synced" flag which would separate synced (sent to the server) operations from the unsynced

Then the following properties should hold:

* The `task undo` should be able to undo unsynced user actions by simply removing all the operations belonging to that user action from the log
* `task undo` for synced operations would not be supported

A constant size requirement could be achieved by trimming the synced parts of the operations log, per user settings (i.e. operations older than a year).

This means that the operations log could be used as a source of history of modifications as well, serving the need that the undo.data serves currently.

**Comment by djmitche** [comment]
Closing this as it's past the point of an RFC now :)
